### DISCUSSION TITLE
Why does no one talk about sqlites method to allow concurrent writer operations via page locking

### DISCUSSION THREAD

**Discussion by erikd234** [pocketbase/pocketbase#5582] [Q&A]
A huge bottlenecks of sqlite is that for write operations we can only write to the WAL then one single writer is writing those to the real db file.

I was reading [expensify scales to 4million qps](https://use.expensify.com/blog/scaling-sqlite-to-4m-qps-on-a-single-server)

And other posts about this on hackernews and bedrockdb 

and im wondering why more people don't talk about sqlites ability to supports concurrent writers via their page-locking branch (and changesets allow for effectively row-level locks).  

I can hardly find anything on the internet about this!

I thought since this project relies on sqlite so much, i was hoping someone would know the answer AND what about the future idea of pocketbase have an option or using this technology to greatly improve its performance benchmarks for writing when someone scales up their app.

What do you think about sqlite page-locking branch? Would it improve write benchmarks? Could it makes its way into pocketbase? 

Edit: I found this page on the documentation for the feature

`https://sqlite.org/src/doc/begin-concurrent/doc/begin_concurrent.md`

I suppose one could, build that branch from source? or install that and use it instead of the release branches?

**Comment by lumosminima** [comment]
IMO, because people who need concurrent writes are already using dbs that are built specifically for that purpose: client/server db's like mysql, postgres etc. Sqlite has a good page on when to use it: [https://www.sqlite.org/whentouse.html](https://www.sqlite.org/whentouse.html). Feature like Begin Concurrent, [bedrock](https://www.sqlite.org/src/timeline?advm=0&n=100&r=bedrock&ss=v) (WAL2+BEGIN CONCURRENT) and the new[ HC-tree](https://sqlite.org/hctree/doc/hctree/doc/hctree/index.html) based sqlite do provide more concurrency but are not in main branch probably because they are not ready. 

I personally see pocketbase as the starting place to build and prototype from as it's extendable with Go and it's simple, other commercial self hosted solutions like supabase [are not as simple](https://news.ycombinator.com/item?id=40111400), this would be enough for any personal or small-medium sized app you can think of but in case you do decide to make a SaaS or startup app that somehow gets million users overnight it would still be simpler to incrementally migrate towards something more scalable because (a) pocketbase is extendable and stack is simple (b) you probably also have million bucks from all those users ;) It's also kind of like comparing french press to a commercial espresso machine the latter needs more maintenance but also serves more people quickly yet you can't beat the simplicity of a french press.

Discussions on HN:
[SQLite Begin Concurrent](https://news.ycombinator.com/item?id=36893193)
[HC-tree is an experimental high-concurrency database back end for SQLite](https://news.ycombinator.com/item?id=34434025)
